### DISCUSSION TITLE
Changing color of SVG icons and text inside input on focus

### DISCUSSION THREAD

**Discussion by daltonsutton** [tailwindlabs/tailwindcss#1772] [Help] [answered]
I'm trying to change the color of the icon and text inside an input when it's focused. I'd like to have the placeholder color be the same color on default and when it's focused have it change to a darker color and when it's not focused, to a lighter color. I've tried a couple of things like focus: and active: but to no avail.

Here is an example of what I have. The html is from Tailwind UI.
![tailwindinput](https://user-images.githubusercontent.com/3005706/82006488-86f32d80-962d-11ea-80bd-e63fada17b8e.gif)

Any help would be appreciated!

**Comment by adamwathan** [comment]
So here's the snippet from Tailwind UI that does work as I'd expect:

```html
<div class="relative">
  <div class="absolute inset-y-0 left-0 pl-3 flex items-center pointer-events-none">
    <svg class="h-5 w-5 text-gray-400" fill="currentColor" viewBox="0 0 20 20">
      <path
        fill-rule="evenodd"
        d="M8 4a4 4 0 100 8 4 4 0 000-8zM2 8a6 6 0 1110.89 3.476l4.817 4.817a1 1 0 01-1.414 1.414l-4.816-4.816A6 6 0 012 8z"
        clip-rule="evenodd"
      />
    </svg>
  </div>
  <input
    id="search"
    class="block w-full pl-10 pr-3 py-2 border border-transparent rounded-md leading-5 bg-gray-700 text-gray-300 placeholder-gray-400 focus:outline-none focus:bg-white focus:text-gray-900 sm:text-sm transition duration-150 ease-in-out"
    placeholder="Search"
    type="search"
  />
</div>
```

![input-focus-color](https://user-images.githubusercontent.com/4323180/82046464-eded0280-967e-11ea-8e1d-d931ffe6700f.gif)

Can you share your code so we can see what the difference is?

**Reply by daltonsutton** [reply]
This is what I currently have. In my tailwind.config.js file, I have:

```
transparent: {
  0: 'rgba(255, 255, 255, 0)',
  50: 'rgba(255, 255, 255, 0.05)',
  100: 'rgba(255, 255, 255, 0.1)',
  200: 'rgba(255, 255, 255, 0.2)',
  300: 'rgba(255, 255, 255, 0.3)',
  400: 'rgba(255, 255, 255, 0.4)',
  500: 'rgba(255, 255, 255, 0.5)',
  600: 'rgba(255, 255, 255, 0.6)',
  700: 'rgba(255, 255, 255, 0.7)',
  800: 'rgba(255, 255, 255, 0.8)',
  900: 'rgba(255, 255, 255, 0.9)',
},
darktransparent: {
  0: 'rgba(0, 0, 0, 0)',
  50: 'rgba(0, 0, 0, 0.05)',
  100: 'rgba(0, 0, 0, 0.1)',
  200: 'rgba(0, 0, 0, 0.2)',
  300: 'rgba(0, 0, 0, 0.3)',
  400: 'rgba(0, 0, 0, 0.4)',
  500: 'rgba(0, 0, 0, 0.5)',
  600: 'rgba(0, 0, 0, 0.6)',
  700: 'rgba(0, 0, 0, 0.7)',
  800: 'rgba(0, 0, 0, 0.8)',
  900: 'rgba(0, 0, 0, 0.9)',
},
```

And in my HTML, I have the following:
```
<div class="relative">
    <div class="absolute inset-y-0 left-0 pl-3 flex items-center pointer-events-none">
        <svg class="h-5 w-5 text-transparent-300 focus:text-darktransparent-900" fill="currentColor" viewBox="0 0 20 20">
            <path fill-rule="evenodd" d="M8 4a4 4 0 100 8 4 4 0 000-8zM2 8a6 6 0 1110.89 3.476l4.817 4.817a1 1 0 01-1.414 1.414l-4.816-4.816A6 6 0 012 8z" clip-rule="evenodd" />
        </svg>
    </div>
    <input id="search" class="block w-full pl-10 pr-3 py-2 rounded-md leading-5 bg-transparent-100 text-transparent-500 placeholder-transparent-400 focus:outline-none focus:bg-white focus:text-darktransparent-900 sm:text-sm transition duration-150 ease-in-out" placeholder="Search" type="search" />
</div>
```

I think the main difference is that in your example, you're using "text-gray-400" for your SVG color, and that's the perfect color for both dark and light backgrounds with the input is in and out of focus. I'm wondering if there is a way to have to change colors on focus for the SVG? I like your trick though and may just go with that, it seems the easiest way. I figured out changing the text color on focus after submitting this last night, just the SVG doesn't seem to like that pseudo class.

Thanks for helping me with this!

**Comment by Apostolique** [comment]
I wonder if it can be done with focus-within: https://tailwindcss.com/docs/pseudo-class-variants/#focus-within

**Comment by estevanmaito** [comment, marked as answer]
It can also be done with `focus-within`:

```html
<div class="relative max-w-xs text-gray-600 focus-within:text-gray-800">
  <div class="absolute inset-y-0 left-0 flex items-center pl-2 pointer-events-none">
    <svg class="w-4 h-4 fill-current" viewBox="0 0 20 20">
      <path
        d="M10 2a6 6 0 00-6 6v3.586l-.707.707A1 1 0 004 14h12a1 1 0 00.707-1.707L16 11.586V8a6 6 0 00-6-6zM10 18a3 3 0 01-3-3h6a3 3 0 01-3 3z"></path>
    </svg>
  </div>
  <input
    class="w-full h-10 pl-8 text-gray-700 placeholder-gray-600 bg-gray-200 rounded-lg shadow-md focus:bg-white"
    type="text"
    placeholder="Search for something" />
</div>
```

![colored-icon](https://user-images.githubusercontent.com/8527573/82089604-4d243480-96ca-11ea-99eb-d0f1fb1bbe35.gif)

In this case I added `focus-within:text-gray-800` to the wrapper `div`.

Also, add `focus-within` to your `tailwind.config.js`

```javascript
variants: {
  textColor: ['responsive', 'hover', 'focus', 'focus-within'],
},
```

**Comment by Apostolique** [comment]
It could be fun to have a utility similar to `group-hover` or `group-focus` for `focus-within`.

```html
<div class="group">
  <div class="text-gray-700 group-focus-within:text-gray-800">
    <svg class="w-4 h-4 fill-current" viewBox="0 0 20 20">
      <path d="M10 2a6 6 0 00-6 6v3.586l-.707.707A1 1 0 004 14h12a1 1 0 00.707-1.707L16 11.586V8a6 6 0 00-6-6zM10 18a3 3 0 01-3-3h6a3 3 0 01-3 3z"></path>
    </svg>
  </div>
  <input
    class="text-gray-700 group-focus-within:text-gray-800"
    type="text"
    placeholder="Search for something" />
</div>
```

**Reply by benface** [reply]
I agree, and it's part of my interaction variants plugin: https://github.com/benface/tailwindcss-interaction-variants

**Reply by Apostolique** [reply]
That's awesome!

**Reply by josineto** [reply]
I agree too! Anyway, thank you, @benface !

**Reply by HugoRamonPereira** [reply]
perfect, solution, thank you so much @Apostolique

**Comment by daltonsutton** [comment]
Thanks guys for your help! I was able to get this working!

**Comment by kripod** [comment]
I would like to ask if there are any plans to support the `group-focus-within` variant out of the box, as mentioned above. With the release of Tailwind v2, [variants can be extended easier](https://tailwindcss.com/docs/configuring-variants#enabling-extra-variants), but custom utilities should be added to [`variantOrder`](https://tailwindcss.com/docs/configuration#variant-order) like below, which is a cumbersome task to do:

```js
/* tailwind.config.js */
module.exports = {
  // […]
    variantOrder: [
    "first",
    "last",
    "odd",
    "even",
    "visited",
    "checked",
    "group-focus-within", // TODO: Remove when Tailwind adds this
    "group-hover",
    "group-focus",
    "focus-within",
    "hover",
    "focus",
    "focus-visible",
    "active",
    "disabled",
  ],
};
```

Thanks for helping in advance!
